src/functional/sim_piece.zig

```diff
@@ -0,0 +1,765 @@
+            var previous_timestamp: usize = 0;
+            // Iterate over the timestamps and append them.
+            for (hash_to_array_entry.value_ptr.*.items) |timestamp| {
+                try appendValue(usize, timestamp - previous_timestamp, compressed_values);
+                previous_timestamp = timestamp;
+            }
+        }
+    }
+}
+
+/// Quantizes the given `value` by the specified `error_bound`. This process ensures that
+/// the quantized value remains within the error bound of the original value. If the
+/// `error_bound` is equal to zero, the value is directly returned.
+fn quantize(value: f64, error_bound: f32) f64 {
+    if (error_bound != 0)
+        return @floor(value / error_bound) * error_bound;
+    return value;
+}
+
+/// Appends the `metadata` to the hash map `metadata_map`. The `allocator` to use for allocating
+/// the memory for a new ArrayList if the `metadata.interception` does not exist.
+fn appendSegmentMetadata(
+    metadata_map: *HashMap(
+        f64,
+        ArrayList(SegmentMetadata),
+        HashF64Context,
+        std.hash_map.default_max_load_percentage,
+    ),
+    metadata: SegmentMetadata,
+    allocator: mem.Allocator,
+) !void {
+    const get_result = try metadata_map.getOrPut(metadata.interception);
+    if (!get_result.found_existing) {
+        get_result.value_ptr.* = ArrayList(SegmentMetadata).init(allocator);
+    }
+    try get_result.value_ptr.*.append(metadata);
+}
+
+/// Returns a comparator function that compares SegmentMetadata by `lower_bound_slope`.
+fn compareMetadataBySlope() fn (void, SegmentMetadata, SegmentMetadata) bool {
+    return struct {
+        pub fn inner(_: void, metadata_one: SegmentMetadata, metadata_two: SegmentMetadata) bool {
+            return metadata_one.lower_bound_slope < metadata_two.lower_bound_slope;
+        }
+    }.inner;
+}
+
+/// Returns a comparator function that compares SegmentMetadata by `start_time`.
+fn compareMetadataByStartTime() fn (void, SegmentMetadata, SegmentMetadata) bool {
+    return struct {
+        pub fn inner(_: void, metadata_one: SegmentMetadata, metadata_two: SegmentMetadata) bool {
+            return metadata_one.start_time < metadata_two.start_time;
+        }
+    }.inner;
+}
+
+/// Append `value` of `type` determined at compile time to `compressed_values`.
+fn appendValue(comptime T: type, value: T, compressed_values: *std.ArrayList(u8)) !void {
+    // Compile-time type check
+    switch (@TypeOf(value)) {
+        f64, usize => {
+            const value_as_bytes: [8]u8 = @bitCast(value);
+            try compressed_values.appendSlice(value_as_bytes[0..]);
+        },
+        f32 => {
+            const value_as_bytes: [4]u8 = @bitCast(value);
+            try compressed_values.appendSlice(value_as_bytes[0..]);
+        },
+        else => @compileError("Unsupported type for append value function"),
+    }
+}
+
+test "f64 context can hash" {
```

Comment:
Since the tests in this file use random data, maybe implement them using `tester.zig` and extend it with functions for generating data with the required structure if such functions does not already exist. Since most of our tests for TerseTS consist of generating data, compress, decompress, and check, implementing functionality for this in a single place instead of in each test seems like a good way to significantly reduce code duplication and the amount of effort required to write tests.

Reply:
This was a good idea. A lot of duplicated code is gone.